### PR TITLE
Unify allocate/delete_device_data in base/cuda.h

### PR DESCRIPTION
**Pull request by masterleinad**
This PR unifies the various `allocate/delete_device_*` we introduced to be able to use `std::unique_ptr` in the `CUDAWrappers` classes.

### CODE REVIEW

File: include/deal.II/base/cuda.h [hunk 1]
```diff
@@ -103,6 +103,39 @@ namespace Utilities
...
+    void
+    delete_device_data(Number *device_ptr) noexcept
+    {
+#ifdef DEAL_II_COMPILER_CUDA_AWARE
+      const cudaError_t error_code = cudaFree(device_ptr);
+      (void)error_code;
+      AssertNothrow(error_code == cudaSuccess,
+                    dealii::ExcCudaError(cudaGetErrorString(error_code)));
+#else
+      (void)device_ptr;
```
Comment (bangerth) [inline include/deal.II/base/cuda.h]: Can you make this an `Assert(false, ...)`? We shouldn't get here, of course, but it's useful to check :-)

File: include/deal.II/base/cuda.h [hunk 2]
```diff
@@ -103,6 +103,39 @@ namespace Utilities
...
+    template <typename Number>
+    Number *
+    allocate_device_data(const std::size_t size)
+    {
+#ifdef DEAL_II_COMPILER_CUDA_AWARE
+      Number *device_ptr;
+      Utilities::CUDA::malloc(device_ptr, size);
+      return device_ptr;
+#else
+      (void)size;
```
Comment (bangerth) [inline include/deal.II/base/cuda.h]: same here

File: include/deal.II/base/cuda.h [hunk 3]
```diff
@@ -103,6 +103,46 @@ namespace Utilities
...
+    /**
+     * Deleter to be used for `std::unique_ptr` pointing to device memory.
+     */
+    template <typename Number>
+    void
+    delete_device_data(Number *device_ptr) noexcept
+    {
+#ifdef DEAL_II_COMPILER_CUDA_AWARE
+      const cudaError_t error_code = cudaFree(device_ptr);
+      (void)error_code;
```
Comment (Rombur) [inline include/deal.II/base/cuda.h]: Not necessary, it should be taken care of in `AssertNothrowCuda`

File: include/deal.II/base/cuda.h (hunk 2 above)
Comment (masterleinad) [reply include/deal.II/base/cuda.h]: Fixed.

File: include/deal.II/base/cuda.h (hunk 3 above)
Comment (masterleinad) [reply include/deal.II/base/cuda.h]: Fixed.

**Comment by masterleinad** [comment]
./run-tests

**Pull request merged**